### PR TITLE
Update DDP Tutorial to remove Single-Process Multi-Device Use Case

### PR DESCRIPTION
**Pull request by mrshenli**
As we do not recommend using Single-Process Multi-Device mode with
DDP due to the reduced throughput and as this mode is broken in
v1.5, added this PR to update DDP tutorial to using Multi-Process
Single-Device mode.

This PR also includes the following minor fixes:

1. highlight that this tutorial requires at least 8 GPUs to run
2. added a pointer to RPC
3. added a pointer to torchelastic
4. as DDP now broadcast initial param values from rank 0 to others
   it no longer require setting RNG seed.

### CODE REVIEW

**Comment by mrshenli** [comment]
The corresponding example PR is https://github.com/pytorch/examples/pull/763

**Comment by mrshenli** [comment]
Hey @kiukchung this currently links to https://github.com/pytorch/elastic. Let me know if you prefer https://pytorch.org/elastic/0.2.0rc1/index.html or a different link. Thanks!

**Comment by mrshenli** [comment]
cc @jlin27 @gchanan @zhaojuanmao @rohan-varma

File: intermediate_source/ddp_tutorial.rst [hunk 1]
```diff
@@ -2,46 +2,53 @@ Getting Started with Distributed Data Parallel
...
-passes and gradient synchronizations in backward passes. It is up to users to
-map processes to available resources, as long as processes do not share GPU
-devices. The recommended (usually fastest) approach is to create a process for
-every module replica, i.e., no module replication within a process. The code in
-this tutorial runs on an 8-GPU server, but it can be easily generalized to
-other environments.
+`DistributedDataParallel <https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel>`__
+(DDP) implements data parallelism at the module level which can run across
+multiple machines. Applications using DDP should spawn multiple processes and
+create a DDP instance in each process. DDP uses collective communications in the
```
Comment (rohan-varma) [inline intermediate_source/ddp_tutorial.rst:8]: nit: "create a single DDP instance per process" might be clearer?

File: intermediate_source/ddp_tutorial.rst [hunk 2]
```diff
@@ -2,46 +2,53 @@ Getting Started with Distributed Data Parallel
...
+`torch.distributed <https://pytorch.org/tutorials/intermediate/dist_tuto.html>`__
+package to synchronize gradients and buffers. More specifically, DDP inserts
+an autograd hook for each model parameter which will fire when the
```
Comment (rohan-varma) [inline intermediate_source/ddp_tutorial.rst:11]: Is the autograd hook indeed per-parameter as given by `model.parameters()`? If user takes a model and then manually adds a `nn.Parameter()` to it, will DDP sync gradients from this parameter as well? If so, might be worth mentioning this (maybe in the docs is better though)

File: intermediate_source/ddp_tutorial.rst [hunk 3]
```diff
@@ -127,23 +125,27 @@ Otherwise, global gradient synchronizes will not make sense.
...
-In DDP, constructor, forward method, and differentiation of the outputs are
-distributed synchronization points. Different processes are expected to reach
-synchronization points in the same order and enter each synchronization point
-at roughly the same time. Otherwise, fast processes might arrive early and
-timeout on waiting for stragglers. Hence, users are responsible for balancing
-workloads distributions across processes. Sometimes, skewed processing speeds
-are inevitable due to, e.g., network delays, resource contentions,
-unpredictable workload spikes. To avoid timeouts in these situations, make
-sure that you pass a sufficiently large ``timeout`` value when calling
+In DDP, the constructor and the backward pass are distributed synchronization
```
Comment (rohan-varma) [inline intermediate_source/ddp_tutorial.rst:139]: Noticed we removed forward method, was it inaccurate that it was a point of synchronization or did something change?

Comment (kiukchung) [reply intermediate_source/ddp_tutorial.rst:139]: > kiuk

@mrshenli that would be awesome, the correct link is https://pytorch.org/elastic (it redirects to the lastest release)

File: intermediate_source/ddp_tutorial.rst (hunk 2 above)
Comment (mrshenli) [reply intermediate_source/ddp_tutorial.rst:11]: If user manually adds a `nn.Parameter()` the `Model.__setattr__` should detect that and call `register_parameter` for it, so this manually added parameter would also be included ini `model.parameters()`. Let me modify this to "each model parameter in `model.parameters()`".

File: intermediate_source/ddp_tutorial.rst (hunk 3 above)
Comment (mrshenli) [reply intermediate_source/ddp_tutorial.rst:139]: > Noticed we removed forward method, was it inaccurate that it was a point of synchronization or did something change?

There is sync in forward to broadcast buffers from rank 0 to all other ranks. I removed this  as 1) there is no intra-node sync in MPSD mode 2) buffer is not used in this tutorial. But give it a second thought, I think it might be better to keep this to stay consistent with the previous version which is correct for general cases. Thanks for spotting this!

Comment (mrshenli) [reply intermediate_source/ddp_tutorial.rst]: links are updated, thx!
